### PR TITLE
Fix OpenGLES 3.0 on Qualcomm's crappy driver, it can't bitshift sometime...

### PR DESCRIPTION
**Pull request by ds84182**
...s.
This commit was tested with a Moto X 2013 on Stock Android Kitkat. I have no clue what other versions this bug is present in, but I know it existed in version 45 of their drivers.

Anyways, what this commit does is add some shader methods for bitshifts if the driver currently has the bug. The functions do scalar by scalar bit shifts (since it magically supports that).

And as always... *thanks Qualcomm*.

### CODE REVIEW

File: Source/Core/VideoCommon/DriverDetails.cpp [hunk 1]
```diff
@@ -49,6 +49,7 @@ namespace DriverDetails
 		{OS_ALL,    VENDOR_QUALCOMM, DRIVER_QUALCOMM,     -1, BUG_BROKENTEXTURESIZE,   -1.0, 65.0, true},
 		{OS_ALL,    VENDOR_QUALCOMM, DRIVER_QUALCOMM,     -1, BUG_BROKENATTRIBUTELESS, -1.0, 94.0, true},
 		{OS_ALL,    VENDOR_QUALCOMM, DRIVER_QUALCOMM,     -1, BUG_BROKENNEGATEDBOOLEAN,-1.0, -1.0, true},
+		{OS_ALL,    VENDOR_QUALCOMM, DRIVER_QUALCOMM,     -1, BUG_BROKENIVECSHIFTS,-1.0, 46.0, true},
```
Comment (lioncash) [inline Source/Core/VideoCommon/DriverDetails.cpp]: Vertical alignment.

File: Source/Core/VideoCommon/LightingShaderGen.h [hunk 2]
```diff
@@ -8,6 +8,7 @@
 #include "VideoCommon/NativeVertexFormat.h"
 #include "VideoCommon/ShaderGenCommon.h"
 #include "VideoCommon/XFMemory.h"
+#include "VideoCommon/DriverDetails.h"
```
Comment (lioncash) [inline Source/Core/VideoCommon/LightingShaderGen.h]: Needs to be alphabetically ordered

File: Source/Core/VideoCommon/PixelShaderGen.cpp [hunk 3]
```diff
@@ -948,12 +1011,37 @@ static inline void WriteTevRegular(T& out, const char* components, int bias, int
...
-	out.Write(" %s ", tevOpTable[op]);
-	out.Write("(((((tevin_a.%s<<8) + (tevin_b.%s-tevin_a.%s)*(tevin_c.%s+(tevin_c.%s>>7)))%s)%s)>>8)",
-	          components, components, components, components, components,
-	          tevScaleTableLeft[shift], tevLerpBias[2*op+(shift!=3)]);
-	out.Write(")%s", tevScaleTableRight[shift]);
+	
+	if (DriverDetails::HasBug(DriverDetails::BUG_BROKENIVECSHIFTS))
+	{
+		//haxx, cleaner code by not having irshift and ilshift in the emitted code by omitting them if not used.
+		const char *leftShift = tevScaleTableLeft[shift];
```
Comment (lioncash) [inline Source/Core/VideoCommon/PixelShaderGen.cpp]: asterisk against the type.

**Comment by ds84182** [comment]
All style issues are fixed now.

**Comment by Mullin** [comment]
LGTM, this is bad for Qualcomm : /

**Comment by magumagu** [comment]
Not particularly happy with this... it's a lot of new code to work around an issue in a driver that doesn't really work very well in the first place.  I don't really have any suggestions to improve it, though.

**Comment by Sonicadvance1** [comment]
Working around video driver bugs is never something to be happy about.
There isn't any way to work around this in a cleaner fashion because it is such a simple thing breaking.
At the very least, it's very well documented and isolated cleanly in to if/else statements.
@dolphin-emu-bot rebuild

File: Source/Core/VideoCommon/DriverDetails.cpp [hunk 4]
```diff
@@ -49,6 +49,7 @@ namespace DriverDetails
 		{OS_ALL,    VENDOR_QUALCOMM, DRIVER_QUALCOMM,     -1, BUG_BROKENTEXTURESIZE,   -1.0, 65.0, true},
 		{OS_ALL,    VENDOR_QUALCOMM, DRIVER_QUALCOMM,     -1, BUG_BROKENATTRIBUTELESS, -1.0, 94.0, true},
 		{OS_ALL,    VENDOR_QUALCOMM, DRIVER_QUALCOMM,     -1, BUG_BROKENNEGATEDBOOLEAN,-1.0, -1.0, true},
+		{OS_ALL,    VENDOR_QUALCOMM, DRIVER_QUALCOMM,     -1, BUG_BROKENIVECSHIFTS    ,-1.0, 46.0, true},
```
Comment (Sonicadvance1) [inline Source/Core/VideoCommon/DriverDetails.cpp]: Comma next to BUG_BROKENIVECSHIFTS, and then spaces to align the numbers.

**Comment by degasus** [comment]
@Sonicadvance1 We still are allowed to say that we don't run on such broken drivers...

@ds84182 You did lots of if(broken) ... else ..., what's wrong with also using this functions on fine drivers? Maybe with a different definitions through

Your definitions are copy&pasted between pixel and vertex shader, maybe we should just use a global shader header else. We do so in ogl right now, so only d3d is missing :/

**Comment by Sonicadvance1** [comment]
If someone is willing to put a bit of time in to supporting the broken drivers I'm fine with it.
Using this solution on every GPU may have a negative impact on SIMD architecture GPUs that have less than ideal video drivers?
Also using it outside of if/else may be better off than embedding it everywhere since it makes it uglier to strip out later when we give up trying to support such old video drivers.

**Comment by ds84182** [comment]
Anything preventing this from getting merged?

**Comment by shuffle2** [comment]
@dolphin-emu-bot rebuild

**Comment by Sonicadvance1** [comment]
Just requires the lint errors to be fixed.

**Comment by Parlane** [comment]
Need to fix lint issues as @Sonicadvance1 has stated, please go to: https://buildbot.dolphin-emu.org/builders/lint/builds/1634/steps/shell_1/logs/stdio

These are all trailing whitespace issues.

Also another thing while you are fixing those is that with comments //, you should put a space after // and use capitals for start of sentences etc.. :\

**Comment by comex** [comment]
In the interests of getting things done, I relinted this myself - #2319

**Pull request closed without merging**